core/src/main/kotlin/org/evomaster/core/search/gene/TupleGene.kt

```diff
@@ -0,0 +1,236 @@
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ *  A tuple is a fixed-size, ordered list of elements, of possible different types.
+ *  This is needed for example when representing the inputs of function calls in
+ *  GraphQL.
+ */
+class TupleGene(
+    /**
+     * The name of this gene
+     */
+    name: String,
+    /**
+     * The actual elements in the array, based on the template. Ie, usually those elements will be clones
+     * of the templated, and then mutated/randomized
+     * note that if the list of gene could be updated, its impact needs to be updated
+     */
+    val elements: List<Gene>,
+    /**
+     * In some cases, we want to treat an element differently from the other (the last in particular).
+     * This is for example the case of function calls in GQL when the return type is an object, on
+     * which we need to select what to retrieve.
+     * In these cases, such return object will be part of the tuple, as the last element.
+     */
+    val lastElementTreatedSpecially: Boolean = false
+
+) : Gene(name, elements) {
+
+    init {
+        if (elements.isEmpty()) {
+            throw IllegalArgumentException("Empty tuple")
+        }
+    }
+
+    companion object {
+        val log: Logger = LoggerFactory.getLogger(TupleGene::class.java)
+    }
+
+    fun getSpecialGene(): Gene? {
+        TODO("Not yet implemented")
+        if (lastElementTreatedSpecially) {
+            return elements.last()
+        }
+        return null
+
+    }
+
+
+    override fun getValueAsPrintableString(
+        previousGenes: List<Gene>,
+        mode: GeneUtils.EscapeMode?,
+        targetFormat: OutputFormat?,
+        extraCheck: Boolean
+    ): String {
+
+        val buffer = StringBuffer()
+
+        //need the name for input and return
+        buffer.append("$name")
+
+        //printout the inputs. See later if a refactoring is needed
+        if (elements.dropLast(1).isNotEmpty()) {
+            buffer.append("(")
+            val s=elements.dropLast(1).map {
+
+                if (it is EnumGene<*> ||
+                    (it is OptionalGene && it.gene is EnumGene<*>) ||
+                    (it is OptionalGene && it.gene is ArrayGene<*> && it.gene.template is EnumGene<*>) ||
+                    (it is OptionalGene && it.gene is ArrayGene<*> && it.gene.template is OptionalGene && it.gene.template.gene is EnumGene<*>) ||
+                    (it is ArrayGene<*> && it.template is EnumGene<*>) ||
+                    (it is ArrayGene<*> && it.template is OptionalGene && it.template.gene is EnumGene<*>)
+                ) {
+                    val i = it.getValueAsRawString()
+                    "${it.name} : $i"
+                } else {
+                    if (it is ObjectGene || (it is OptionalGene && it.gene is ObjectGene)) {
+                        val i = it.getValueAsPrintableString(mode = GeneUtils.EscapeMode.GQL_INPUT_MODE)
```

Comment:
need to consider this gene might be used in non GQL contexts

Reply:
Done